drivers/i3c/i3c_stm32.c

```diff
@@ -0,0 +1,2129 @@
+		data->pid >>= 16;
+
+		/* Find the device in the device list */
+		ret = i3c_dev_list_daa_addr_helper(&data->drv_data.attached_dev.addr_slots,
+						   &config->drv_cfg.dev_list, data->pid, false,
+						   false, &target, &dyn_addr);
+		if (ret != 0) {
+			/* TODO: figure out what is the correct sequence to exit form this error
+			 * It is expected that a TX overrun error to occur which triggers err isr
+			 */
+			LOG_ERR("No dynamic address could be assigned to target");
+
+			return;
+		}
+
+		/* Put the new dynamic address in TX FIFO for transmission */
+		LL_I3C_TransmitData8(i3c, dyn_addr);
+
+		if (target != NULL) {
+			/* Update target descriptor */
+			target->dynamic_addr = dyn_addr;
+			target->bcr = bcr;
+			target->dcr = dcr;
+		}
+
+		/* Mark the address as used */
+		i3c_addr_slots_mark_i3c(&data->drv_data.attached_dev.addr_slots, dyn_addr);
+
+		/* Mark the static address as free */
+		if ((target != NULL) && (target->static_addr != 0) &&
+		    (dyn_addr != target->static_addr)) {
+			i3c_addr_slots_mark_free(&data->drv_data.attached_dev.addr_slots, dyn_addr);
+		}
+
+		break;
+	}
+	case STM32_I3C_MSG_CCC: {
+		struct i3c_ccc_payload *payload = data->ccc_payload;
+
+		if (payload->ccc.num_xfer < payload->ccc.data_len) {
+			LL_I3C_TransmitData8(i3c, payload->ccc.data[payload->ccc.num_xfer++]);
+		}
+		break;
+	}
+	case STM32_I3C_MSG_CCC_P2: {
+		struct i3c_ccc_target_payload *target = data->ccc_target_payload;
+
+		if (target->num_xfer < target->data_len) {
+			LL_I3C_TransmitData8(i3c, target->data[target->num_xfer++]);
+
+			/* After sending all bytes for current target, move on to the next target */
+			if (target->num_xfer == target->data_len) {
+				data->ccc_target_payload++;
+			}
+		}
+		break;
+	}
+	default:
+		break;
+	}
+}
+
+static void i3c_stm32_event_isr_rx(const struct device *dev)
+{
+	const struct i3c_stm32_config *config = dev->config;
+	struct i3c_stm32_data *data = dev->data;
+	I3C_TypeDef *i3c = config->i3c;
+
+	switch (data->msg_state) {
+	case STM32_I3C_MSG: {
+		uint8_t *buf = NULL;
+		size_t *offset = NULL;
+		uint32_t len = 0;
+
+		i3c_stm32_curr_msg_xfer_get_buf(dev, &buf, &len, &offset);
+		if (i3c_stm32_drain_rx_fifo(dev, buf, len, offset)) {
+			i3c_stm32_curr_msg_xfer_next(dev);
+		}
+
+		break;
+	}
+	case STM32_I3C_MSG_DAA: {
+		data->pid <<= 8;
+		data->pid |= LL_I3C_ReceiveData8(i3c);
+
+		data->daa_rx_rcv++;
+
+		/* After receiving 8 PID bytes from DAA, enable TXFNF interrupt to send the dynamic
+		 * address
+		 */
+		if (data->daa_rx_rcv == 8) {
+			LL_I3C_EnableIT_TXFNF(i3c);
+			data->daa_rx_rcv = 0;
+		}
+		break;
+	}
+	case STM32_I3C_MSG_CCC_P2: {
+		struct i3c_ccc_target_payload *target = data->ccc_target_payload;
+
+		if (target->num_xfer < target->data_len) {
+			target->data[target->num_xfer++] = LL_I3C_ReceiveData8(i3c);
+
+			/* After receiving all bytes for current target, move on to the next target
+			 */
+			if (target->num_xfer == target->data_len) {
```

Comment:
A target can send an EoD when doing a read with a CCC and it's possible for the num_xfer to be less than the data_len. This can happen with variable length CCCs such as GETCAPS and GETMXDS. Does this handle this situation?

Reply:
No it is not handled, we are currently thinking of a way to handle this issue. The problem is that the information about the target early read termination can be obtained from the status FIFO, however, we might be receiving data from the next target before the data from the status FIFO is ready.

One suggestion we are thinking of is to disable the status FIFO for CCC requests and rely on the RXTGTENDF interrupt to detect early target termination and advance the target_ccc_payload pointer.

Reply:
do you plan to address this as apart of this PR?